diff --git a/setup-tdx-config b/setup-tdx-config
index 38ed3c1..e65a53d 100644
--- a/setup-tdx-config
+++ b/setup-tdx-config
@@ -35,10 +35,10 @@ TDX_PPA_ATTESTATION=tdx-attestation-release
 # Enable the installation of DCAP packages from Canonical's
 # repository. This flag is considered during host OS and guest
 # OS setup.
-# Set to 1 to enable
+# Enable installation of attestation packages from Canonical.
 # By default, the attestation components are not installed
 ################################################################
-TDX_SETUP_ATTESTATION=0
+TDX_SETUP_ATTESTATION=1
 
 ################################################################
 # HOST                                                         #
